src/ghga_connector/core/downloading/batch_processing.py

```diff
@@ -129,167 +129,108 @@ def handle_response(self, *, response: str):
-            if isinstance(url_response, URLResponse):
-                self.staging_state.add_staged(file_id=file_id)
-            else:
-                self.staging_state.add_unstaged(file_id=file_id)
-
-        if unknown_ids:
-            self._handle_unknown(unknown_ids)
-
-    def file_ids_remain(self):
-        """Returns if any staged or unstaged file ids remain"""
-        return any((self.staging_state.staged_files, self.staging_state.unstaged_files))
-
-    def get_staged(self):
-        """Return currently staged file ids"""
-        return self.staging_state.staged_files
-
-    def update_staged_files(self):
-        """Delegate updating file_ids for staging and handle wait/retries."""
-        while self.staging_state.update_staged_files_wait(
-            max_wait_time=self.staging_parameters.max_wait_time,
-            work_package_accessor=self.work_package_accessor,
-        ):
-            self.message_display.display(
-                f"No staged files available, retrying in {self.staging_parameters.retry_after}"
-                + f" seconds for {len(self.staging_state.unstaged_files)} unstaged file(s)."
-            )
-            sleep(self.staging_parameters.retry_after)
-
-    def _handle_unknown(self, unknown_ids: list[str]):
-        """Process user interaction for unknown file IDs"""
-        message = (
-            f"No download exists for the following file IDs: {', '.join(unknown_ids)}"
-        )
+        except exceptions.BadResponseCodeError as error:
+            if error.response_code != 404:
+                raise
+            response = None
+        if isinstance(response, URLResponse):
+            del self.unstaged[file_id]
+            self.staged[file_id] = response
+        elif isinstance(response, RetryResponse):
+            self.unstaged[file_id] = time() + response.retry_after
+        else:
+            self.failed.append(file_id)
+
+    def _check_timeout(self):
+        if time() - self.time_started >= self.max_wait_time:
+            raise exceptions.MaxWaitTimeExceededError(max_wait_time=self.max_wait_time)
+
+    def _handle_failures(self) -> bool:
+        """Handle failed downloads and return whether there was user interaction."""
+        if not self.failed or self.ignore_failed:
+            return False
+        failed = ", ".join(self.failed)
+        message = f"No download exists for the following file IDs: {failed}"
         self.message_display.failure(message)
-
+        if self.finished:
+            raise exceptions.AbortBatchProcessError()
         unknown_ids_present = (
             "Some of the provided file IDs cannot be downloaded."
             + "\nDo you want to proceed ?\n[Yes][No]\n"
         )
         response = self.io_handler.get_input(message=unknown_ids_present)
         self.io_handler.handle_response(response=response)
         self.message_display.display("Downloading remaining files")
+        self.start_time = time()
```

Comment:
Could it be `time_started`?

Reply:
Yes, thanks for the heads up. Fixed in 55ee6c0.

The idea is that the timer should be restarted after the user interaction since the user may take some time to respond.